### PR TITLE
feat(table-toolbar-search): add onBlur/onFocus handling

### LINKED ISSUE #10077: [TableToolbarSearch]: handle passed onFocus/onBlur events without breaking expand behavior
### Summary

We would like to trigger a search callback to update the table data when the search field is blurred, however, this breaks the automatic expand/contract behavior of the search input when either of these events are passed. It would be nice if the internal onFocus/onBlur would perform their actions and then call the external onFocus/onBlur callbacks afterward.

Example reproduction:
https://codesandbox.io/s/weathered-feather-mh35x?file=/src/index.js

1. Click on the Search icon
2. click outside of the search box

Expected: The search box should collapse

### Justification

_No response_

### Desired UX and success metrics

_No response_

### Required functionality

_No response_

### Specific timeline issues / requests

_No response_

### Available extra resources

Yes, I will happily write the PR if this behavior is acceptable.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #10077

add handling to onBlur/onFocus props passed to the TableToolbarSearch. If they are provided, they are called with the event and the handleExpand function as parameters. It's then up to the consumer to call handleExpand to expand/contract the input as desired. Existing automatic expand/contract is maintained if no onBlur/onFocus events are passed.

#### Changelog

**New**

- onBlur prop supported for TableToolbarSearch
- onFocus prop supported for TableToolbarSearch
- added tests to confirm

#### Testing / Reviewing

- added onFocus/onBlur actions to the data-table with toolbar story.
- unit tests should pass

### CODE REVIEW

File: packages/react/src/components/DataTable/TableToolbarSearch.js [hunk 1]
```diff
@@ -96,6 +98,9 @@ const TableToolbarSearch = ({
     }
   };
 
+  const internalOnFocus = (event) => handleExpand(event, true);
```
Comment (abbeyhrt) [inline packages/react/src/components/DataTable/TableToolbarSearch.js]:
```suggestion
  const handleOnFocus = (event) => handleExpand(event, true);
```

We are working towards a standard of internal functions using `handleOn...()` with the prop for that functionality being `on...()`, like you've already done here :)

File: packages/react/src/components/DataTable/TableToolbarSearch.js [hunk 2]
```diff
@@ -96,6 +98,9 @@ const TableToolbarSearch = ({
...
+  const internalOnBlur = (event) => !value && handleExpand(event, false);
```
Comment (abbeyhrt) [inline packages/react/src/components/DataTable/TableToolbarSearch.js]:
```suggestion
  const handleOnBlur = (event) => !value && handleExpand(event, false);
```

Same as above

File: packages/react/src/components/DataTable/TableToolbarSearch.js [hunk 3]
```diff
@@ -110,8 +115,10 @@ const TableToolbarSearch = ({
       }
       onChange={onChange}
       onClear={onClear}
-      onFocus={(event) => handleExpand(event, true)}
-      onBlur={(event) => !value && handleExpand(event, false)}
+      onFocus={
+        onFocus ? (event) => onFocus(event, handleExpand) : internalOnFocus
```
Comment (abbeyhrt) [inline packages/react/src/components/DataTable/TableToolbarSearch.js]:
```suggestion
        onFocus ? (event) => onFocus(event, handleExpand) : handleOnFocus
```

File: packages/react/src/components/DataTable/TableToolbarSearch.js [hunk 4]
```diff
@@ -110,8 +115,10 @@ const TableToolbarSearch = ({
...
+      }
+      onBlur={onBlur ? (event) => onBlur(event, handleExpand) : internalOnBlur}
```
Comment (abbeyhrt) [inline packages/react/src/components/DataTable/TableToolbarSearch.js]:
```suggestion
      onBlur={onBlur ? (event) => onBlur(event, handleExpand) : handleOnBlur}
```

**Review by abbeyhrt — CHANGES_REQUESTED** [review]
Looks good! Just a couple small changes to fit with our code practices

File: packages/react/src/components/DataTable/TableToolbarSearch.js (hunk 1 above)
Comment (kevinsperrine) [reply packages/react/src/components/DataTable/TableToolbarSearch.js]: Thanks, @abbeyhrt. Updated.

**Review by abbeyhrt — APPROVED** [review]
Looks good to me! Thanks for contributing!

**Review by jnm2377 — CHANGES_REQUESTED** [review]
When I look at the `with Toolbar` story, the search is no longer expanding 😓

**Comment by kevinsperrine** [comment]
> When I look at the `with Toolbar` story, the search is no longer expanding 😓

Oh, weird! When I added the bit to include the arguments in the story action, it breaks and causes a cross-origin error in the deploy preview. I'll remove that bit and push it up.

**Review by jnm2377 — APPROVED** [review]

**Comment by jnm2377** [comment]
Thanks for the contribution! 🎉

**Pull request merged**